### DISCUSSION TITLE
Tracker and TODOs for Mill 0.11

### DISCUSSION THREAD

**Discussion by lefou** [com-lihaoyi/mill#1826] [Development]
This thread is meant to collect and track all changes we want to apply before we release Mill 0.11.
We already postphoned some changes to avoid binary compatibility issue. Ideally, we can collect them here, so we don't forget them before the next major release.

**Comment by lefou** [comment]
* [x] make `MultiLogger` an non-case class, and make `debugEnabled` an ctr property

**Comment by lefou** [comment]
* [x] https://github.com/com-lihaoyi/mill/pull/1811

**Comment by lefou** [comment]
* [x] Review all places where `@deprectated` annotations have a `since` before Mill 0.10.0 and remove then when appropriate

**Reply by lefou** [reply]
* [x] Merge https://github.com/com-lihaoyi/mill/pull/2031

**Reply by lolgab** [reply]
Do we need to keep post 0.10 deprecations? We can kindly ask our users to upgrade to the latest 0.10 and fix all the deprecations before upgrading to 0.11.
This would allow us to cleanup a lot of garbage.

**Reply by lefou** [reply]
The thing is, Mill is currently not reporting any deprecated API. You only see them (potentially), when you use a IDE to edit the `build.sc`. This means, the user experience with a 0.11 release could become very frustrating. My current tendency is to keep the stuff that was deprecated after 0.10.0 for the 0.11.x round. I'm open to re-think it, based on good counter examples/arguments.

**Reply by lolgab** [reply]
Maybe what we need then is to report the deprecated apis so we can help users get rid of them. If we wait for 0.12 we are just delaying the pain to later. Having our (already pretty good) deprecation messages on the cli would give users a way to update their builds and save theirselves from future pain.

**Reply by lefou** [reply]
Is this so much of a pain? Dropping API in favor of some, that wasn't even available in 0.10.0 makes it rather hard for plugin authors to support multiple versions. (Not impossible, but also not trivial) So, we probably should only drop them, if we have a really good argument for it. The minimal precondition is to enable these deprecation messages as soon as possible in a next 0.10.x release. And I haven't evaluated how easily these messages can be enabled with our current Ammonite-based setup.

**Reply by lefou** [reply]
@lolgab I think you mean the rather large deprecated set in `scalajslib`. I agree, that we can clean that one more drastically. Scala.js is still such a fast moving target, and users will probably accept, that we can't go with all the newer features without sacrifices. Do you mind to create e PR with the deprecated removal of the old worker? I also think, we forgot to deprecate the `fastOpt` and `fullOpt` targets in favor of the new `fastLinkJs` and `fullLinkJs` targets. We can still deprecate them in `0.10.11`.

**Reply by lolgab** [reply]
I'm also investigating if we can use Ammonite to report deprecation warnings on compilation with something like [this](https://github.com/com-lihaoyi/Ammonite/blob/0f0d597f04e62e86cbf76d3bd16deb6965331470/amm/src/test/resources/scriptCompilerSettings/yRangepos.sc#L4).
It would be also nice if we could update the macros to print deprecation messages when you execute a deprecated target from the cli.

**Reply by lefou** [reply]
We might want to integrate the deprecation info into our `Scaladoc` annotation. Then, the user can see it with `mill inspect`.

**Reply by lefou** [reply]
* [ ] https://github.com/com-lihaoyi/mill-moduledefs/issues/4

**Reply by lefou** [reply]
* [x] #2324

**Comment by lefou** [comment]
* [x] https://github.com/com-lihaoyi/mill/pull/2025

**Comment by lefou** [comment]
* [x] Review dependencies
  * [x] utest >=0.8.0 was previously discarded because it dropped support for Scala.js 0.6

**Reply by lefou** [reply]
* https://github.com/com-lihaoyi/mill/pull/2253

**Comment by lefou** [comment]
* [x] Drop support for Scala.js 0.6

**Reply by lefou** [reply]
@lolgab Can you send a PR?

**Reply by lefou** [reply]
I started a PR but one tests seems to work differently when run with Scala.JS 1

* https://github.com/com-lihaoyi/mill/pull/2218

**Comment by lefou** [comment]
Refactor the command line options configuration class.

* [x] Do not inherit from Ammonite config(s)
* [x] Do not use a case class, but a simple class, to be easily extensible in a bin-compat way
* [x] Review options and drop unsupported options 
* [x] Review #1960

**Comment by lefou** [comment]
~Consider releasing most mill artifacts with the mill platform, e.g. `mill-scalalib_mill0.11_2.13`~

**Reply by lolgab** [reply]
What is the reasoning behind this?

**Reply by lefou** [reply]
The reasoning is, we want to make sure they work within the same Mill ABI. This is probably redundant, as long as we ship all mill packages with each release and include them in the classpath by default. But someday Mill core might be more stable than the specific modules, so we could decouple release cycles, which makes the extra API version necessary. We could also stop bringing them all in the classpath by-default. Once, we want them to load, we could use the same loading mechanism as for external plugins.

On the other side, most ABI is defined by the specific modules, and we most likely will never decouple them from core. I'm unsure. It was just a cause for thought, I guess.

**Comment by lolgab** [comment]
Consider creating `package mill.javalib` to contain `JavaModule` and other Java specific code.

**Reply by lefou** [reply]
What is the issue with having it in package `mill.scalalib`? Just the name? Do you need to depend on `javalib` but conflict with something in `scalalib`?

**Reply by lolgab** [reply]
No, it's just the name. I was thinking that Mill could be a great Java build tool. It would be nice if a Java build didn't mention the Scala language at all.
It sounds maybe stupid since the build itself is written in Scala. But for example Gradle uses Groovy (and lately Kotlin) for the build and is one of the most loved Java build tools.

**Reply by lefou** [reply]
Haha, yeah, Mill already _is_ a great Java build tool. But it uses Zinc under the hood, to compile Java code, so we are using the typical Scala tool chain to build. And placing the Zinc worker into a `javalib`, with all it's Scala intricacies, wouldn't be that ideal either.

> But for example Gradle [...] is one of the most loved Java build tools.

Also, I tend to disagree. Gradle has a large user base, of course, but only a small part of it is loving it. Gradle is really hard to grasp, esp. when things go wrong. There is no static mental model one can build or inspect.

I think a unique selling point of Mill it the easy correspondence to the host language and how easy it is to find the implementation for anything in the build model.

**Reply by lefou** [reply]
What we probably want is a landing page for Java developers or those looking for a good Java build tool. Beside some examples, we could show some comparisons for those projects in different build tools, so those users who have used one of the others before, get a quick idea how it may feel like to use Mill.

**Reply by lolgab** [reply]
> What we probably want is a landing page for Java developers or those looking for a good Java build tool. Beside some examples, we could show some comparisons for those projects in different build tools, so those users who have used one of the others before, get a quick idea how it may feel like to use Mill.

This is a great idea!

**Reply by lolgab** [reply]
Would you consider a `package object` with type aliases for JavaModule and other java specific code under `mill.javalib`?
It would be nice to have no mention of Scala in java projects :)

**Reply by lefou** [reply]
Ugh, that is for sure a little maintenance hell. I'd like to avoid it, but on the other side, I'm not sure how many entries it will have. The greatest issue I see is, that this has potential for great confusion. You could provide a POC though, but I'm not convinced. I agree it would be nice to have pure java stuff in a javalib, but this would mean, we produce lots of source incompatibilities.

**Comment by lolgab** [comment]
- [x] Make ScalaNativeModule use `Seq` instead of `Array` for user exposed targets.
  https://github.com/com-lihaoyi/mill/pull/2193

`ScalaNativeModule` used `Array[String]` for things like `nativeCompileOptions` or `nativeLinkingOptions` because the worker was written in Scala 2.12. Actually it still is using Scala 2.12 for Scala Native `0.4.[0,1,2]` but we already did separate the worker internal API from the user facing API in ScalaJSModule and it worked out pretty well. We can do the same for Scala Native so we can expose more idiomatic types like `Seq[String]`

**Reply by lolgab** [reply]
I want to propose to drop support for Scala Native `0.4.[0,1,2]`. This will simplify the implementation, also I don't think anybody still uses Scala Native `0.4.2` now that there is `0.4.9`. If we do I can convert the worker API from Java to Scala, use Scala types directly and reduce the conversions. Also the implementation will look more similar to the Scala.js one.

**Reply by lefou** [reply]
Yeah, sounds reasonable.

**Comment by lefou** [comment]
* [x] Move `mill.module.Jvm.JarManifest` to `mill.api.JarManifest`, but keep some forwarders (`val` for `object`, `type` for `class`) to keep source compatibility. Also remove the `case` from `class`.

**Reply by lefou** [reply]
* https://github.com/com-lihaoyi/mill/pull/2325

**Comment by lefou** [comment]
The `main` branch is now open for more fundamental changes, including binary breakers. We now  publish snapshot-releases with the `0.11.0-M0` version and will increase the counter after the `M` (which stands for Milestone) when needed.

Now is the time to update old open PRs and submit new with code cleanups and changes. Feel free to use this tracker discussion to link to ideas or issues, which we should include in Mill `0.11`.

**Comment by lefou** [comment]
* [x] Implement a script runner, that can run a pre-compiled Mill buildfile without Ammonite

**Reply by lefou** [reply]
Not a blocker for 0.11.

**Reply by lefou** [reply]
* [x] Realized in https://github.com/com-lihaoyi/mill/pull/2377

**Comment by lefou** [comment]
* [x] Re-integrate `--watch` support back into Mill. (We migrated it into Ammonite for Mill `0.10.0`.)

**Reply by lefou** [reply]
Not a blocker for 0.11.

**Reply by lefou** [reply]
This will be a part of PR

* https://github.com/com-lihaoyi/mill/pull/2377

**Comment by lefou** [comment]
* [x] Do not change the `millSourcePath` in `Cross` modules by default
  * https://github.com/com-lihaoyi/mill/discussions/2252

**Reply by lefou** [reply]
* [x] https://github.com/com-lihaoyi/mill/pull/2347

**Comment by lefou** [comment]
* [ ]  Write a migration / upgrade guide

Topics below:

**Reply by lefou** [reply]
* [ ] Moved `JarManifest` class

**Reply by lefou** [reply]
* [x] Changed default for `millSourcePath` in `Cross` modules

**Reply by lefou** [reply]
* [ ] Changed signatures (TODO: review and list them here)

**Reply by lefou** [reply]
* [ ] Changes / improvements in inter-module resolution but API changes in `recursiveIvyDeps` and `trnsitiveIvyDeps`, which now use `BoundDep` instead of `Dep`
* [ ] Changes in `CoursierModule.resolveDeps` which now wants `BoundDep`s.

**Reply by lefou** [reply]
* [ ] Review #2377 for documentation updates and migration information

**Reply by lefou** [reply]
* [ ] Review https://github.com/com-lihaoyi/mill/pull/2402

**Comment by lefou** [comment]
* [ ] https://github.com/com-lihaoyi/mill/discussions/1779

Probably unrealistic, to get this one into 0.11 as it's a harder task, but very useful. It will definitely break compatibility.

**Comment by lefou** [comment]
* [x] Move `mill.modules.Assembly` to `mill.scalalib.Assembly` and keep source-compat forwarders.

**Reply by lefou** [reply]
* See https://github.com/com-lihaoyi/mill/pull/2513

**Comment by lefou** [comment]
* [x] Unify MillBuildModule and MillBuildRootModule
  * https://github.com/com-lihaoyi/mill/issues/2414
  * https://github.com/com-lihaoyi/mill/pull/2415

**Comment by lihaoyi** [comment]
@lefou what do you think about tagging `0.11.0-M8` and re-bootstrapping the com-lihaoyi/mill repo on top of it soon? I would like to begin dogfooding the changes, especially https://github.com/com-lihaoyi/mill/pull/2377, to try and surface any issues that may not get caught by our test suite. It would also let us clean up our `build.sc` and reset `mill-bootstrap.patch`, which is growing a bit large (especially once https://github.com/com-lihaoyi/mill/pull/2425 lands)

* [x] Release Mill 0.11.0-M8
* [ ] Re-Bootstrap Mill build based on Mill 0.11.0-M8

**Reply by lefou** [reply]
That is the plan, but I wanted to merge all known regressions first.

* [x] #2420

**Reply by lihaoyi** [reply]
Sounds good. Looking forward

**Reply by lefou** [reply]
* [x] Prepare changelog

**Comment by lefou** [comment]
* [x] Restore the ability to run Mill with custom repositories which require authentication (we removed `predef` script support)

**Comment by lihaoyi** [comment]
@lefou seems Coursier now supports `COURSIER_CREDENTIALS` as an environment variable and a `~/.config/coursier/credentials.properties` config file

- https://get-coursier.io/docs/other-credentials

Maybe we can just tell people to use that when they have custom repositories that need auth?

**Reply by lefou** [reply]
Would be nice, if someone could verify, that this really works. There are other issue with configuring coursier, e.g. 

* https://github.com/com-lihaoyi/mill/issues/2312

so we should check it and explicitly document, what's supposed to work and what not.

**Reply by lihaoyi** [reply]
I did some local testing using reposilite; it appears to be successful at redirecting to a local mirror and using credentials:

```
lihaoyi 1-simple-scala$ COURSIER_CREDENTIALS="localhost test-name:tst-secret" COURSIER_REPOSITORIES="http://localhost:8080/private" /Users/lihaoyi/.cache/mill/download/0.11.0-M8-34-19ec00-DIRTY56d084cd -i foo.compile
[build.sc] [4/49] mill.scalalib.ZincWorkerModule.classpath
1 targets failed
mill.scalalib.ZincWorkerModule.classpath
Resolution failed for 1 modules:
--------------------------------------------
  com.lihaoyi:mill-scalalib-worker_2.13:0.11.0-M8-34-19ec00-DIRTY56d084cd
	unauthorized: http://localhost:8080/private/com/lihaoyi/mill-scalalib-worker_2.13/0.11.0-M8-34-19ec00-DIRTY56d084cd/mill-scalalib-worker_2.13-0.11.0-M8-34-19ec00-DIRTY56d084cd.pom (Reposilite)

--------------------------------------------

For additional information on library dependencies, see the docs at
https://com-lihaoyi.github.io/mill/mill/Library_Dependencies.html

lihaoyi 1-simple-scala$ COURSIER_CREDENTIALS="localhost test-name:test-secret" COURSIER_REPOSITORIES="http://localhost:8080/private" /Users/lihaoyi/.cache/mill/download/0.11.0-M8-34-19ec00-DIRTY56d084cd -i foo.compile
[build.sc] [4/49] mill.scalalib.ZincWorkerModule.classpath
1 targets failed
mill.scalalib.ZincWorkerModule.classpath
Resolution failed for 1 modules:
--------------------------------------------
  com.lihaoyi:mill-scalalib-worker_2.13:0.11.0-M8-34-19ec00-DIRTY56d084cd
	not found: http://localhost:8080/private/com/lihaoyi/mill-scalalib-worker_2.13/0.11.0-M8-34-19ec00-DIRTY56d084cd/mill-scalalib-worker_2.13-0.11.0-M8-34-19ec00-DIRTY56d084cd.pom

--------------------------------------------

For additional information on library dependencies, see the docs at
https://com-lihaoyi.github.io/mill/mill/Library_Dependencies.html
```

Will need to try again post-release to verify it actually can download, but I don't have any reason to believe it won't. Given the change in error message from `unauthorized` to `not found`, it's clear that the two environment variables are taking effect

**Comment by lefou** [comment]
* [x] https://github.com/com-lihaoyi/mill/issues/2461

**Comment by lihaoyi** [comment]
@lefou I think I've gotten in most of the major changes I wanted to get in for 0.11.0. Shall we tag 0.11.0-M9? Then we can rebootstrap, roll it out across com-lihaoyi, and let it bake a bit before 0.11.0 final

**Reply by lefou** [reply]
This is the way. 😉

**Reply by lihaoyi** [reply]
Ok, I pushed tag 0.11.0-M9

**Reply by lihaoyi** [reply]
@lolgab @lefou 0.11.0-M9 should be out, if you can help publish mill-mima and mill-vcs- version we can rebootstrap

**Reply by lolgab** [reply]
`mill-mima` `0.0.20` was released with support for Mill `0.11.0-M9`

**Comment by lihaoyi** [comment]
@lefou @lolgab I tagged 0.11.0-M10, with another set of fixes for rough edges I discovered when trying to roll out 0.11.0-M9 throughout com-lihaoyi. If you could publish mill-mima and mill-vcs-version, then I'll proceed with updating the rest of com-lihaoyi to 0.11.0-M10

**Reply by lefou** [reply]
* [x] https://github.com/lefou/mill-vcs-version/pull/112 => Version `0.3.1-8-37c08a`

**Reply by lolgab** [reply]
- [x] https://github.com/lolgab/mill-mima/pull/109 => Version `0.0.21`

**Comment by DamianReeves** [comment]
I've been playing with `0.11.0-M9` and I like the changes. I know @lolgab mentioned it in the PR for the cross Platform stuff, but I want to reiterate the `enable` functionality like what `mill-cross-platform` provides would be really beneficial.

**Reply by lihaoyi** [reply]
I agree, I'd like to get that in somehow. Last I looked I couldn't figure out a way to make it fit nicely, but maybe with some more thought I can find a way

**Comment by lihaoyi** [comment]
Last few PRs for 0.11.0 appear to be the ones I merged below

1. https://github.com/com-lihaoyi/mill/pull/2550
2. https://github.com/com-lihaoyi/mill/pull/2561
3. https://github.com/com-lihaoyi/mill/pull/2562

I'll also try to get a PR in to upload a doc-site bundle, for the versioned-documentation support in 0.11+.

One thing that is worth reciewing before the release is the packages for which we provide compatibility guarantees for. I listed them out in https://github.com/com-lihaoyi/mill#compatibility-stability, rather than have them be implicitly derived from the build configuration. If we want to take bincompat seriously, we should be explicit what we are (and are not) promising.

I'm going to tag 0.11.0-M11 on `f944ceb96565385953ba7fe6869100da24025b03  Sun Jun 4 06:43:06 2023 -0700 Update thirdparty tests, make subprocess testrunner run user code in root classloader (#2561)`, and ask that @lefou and @lolgab release mill-vcs-version and mill-mima again for that milestone. I will then update all the mill-upgrade PRs for the other com.lihaoyi libraries to use the latest code before we tag 0.11.0 final, giving me a chance to catch any last-minute issues as well as helping me draft a migration guide given the com-lihaoyi libraries as examples. Especially the removal of `Tests` traits will require some updates to the existing mill-upgrade PRs

Still aiming for 0.11.0 final to be released on Tuesday morning PST (6 June), so I can advertise it at Scaladays Seattle (7 June), as discussed in Discord.

**Reply by lefou** [reply]
mill-vcs-version version `0.3.1-14-7e2bd2` supports Mill 0.11.0-M11.

**Reply by lolgab** [reply]
mill-mima version `0.0.22` supports Mill `0.11.0-M11`

**Reply by lihaoyi** [reply]
Thanks! I'll update the open mill-upgrade PRs to use M11

**Comment by lihaoyi** [comment]
Merged the last few PRs and tagged 0.11.0. 

* https://github.com/com-lihaoyi/mill/pull/2418
* https://github.com/com-lihaoyi/mill/pull/2575
* https://github.com/com-lihaoyi/mill/pull/2574
* https://github.com/com-lihaoyi/mill/pull/2572

If no flakiness, it should go out in a few hours after it makes its way through CI

**Comment by lihaoyi** [comment]
@lefou @lolgab I started writing `Major changes that may require migration` notes in the release https://github.com/com-lihaoyi/mill/releases/tag/0.11.0. Please help flesh it out with anything I may have missed, and we can use that as the source of truth for the migration notes.

What do you guys think of moving the changelog out of `readme.adoc` into the Github Releases pages? That's what most projects seem to be doing nowadays, and it may be more familiar to users and give them a single page we can link to for all release-related information

**Reply by lefou** [reply]
Well, as a user GitHub Release Pages are nice. From a long time maintenance perspective, I prefer a document in the repository. I'm doing Open Source for 20+ years and I've seen many hosting platforms come an go. I also migrated many project over different platform. Keeping essential info in the repo makes those migrations much easier and older/historic information for users more accessible. Currently GitHub is free to use, but nobody can guarantee it.

The nice thing is, that we can link from the release pages to a changelog or readme file, and even Scala Steward nowadays supports dedicated changelog documents. Also, we can easily render the changelog as part of our documentation.
